segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -263,56 +208,40 @@ private Append getNextAppend() {
         }
     }
 
-    private void handleAppendResult(final Append append, Long newWriteOffset, Throwable exception, Timer elapsedTimer) {
+    private void handleAppendResult(final Append append, Long newWriteOffset, Long previousEventNumber, Throwable exception, Timer elapsedTimer) {
         boolean success = exception == null;
         try {
             boolean conditionalFailed = !success && (Exceptions.unwrap(exception) instanceof BadOffsetException);
-            long previousEventNumber;
-            synchronized (lock) {
-                previousEventNumber = latestEventNumbers.get(Pair.of(append.getSegment(), append.getWriterId()));
-                Preconditions.checkState(outstandingAppend == append,
-                        "Synchronization error in: %s while processing append: %s.",
-                        AppendProcessor.this.getClass().getName(), append);
-            }
 
+            WriterState state = this.writerStates.getOrDefault(Pair.of(append.getSegment(), append.getWriterId()), null);
             if (success) {
-                final DataAppended dataAppendedAck = new DataAppended(append.getRequestId(), append.getWriterId(), append.getEventNumber(),
-                        previousEventNumber, newWriteOffset);
-                log.trace("Sending DataAppended : {}", dataAppendedAck);
-                connection.send(dataAppendedAck);
+                Preconditions.checkState(state != null, "Synchronization error while processing append: %s. Unable to send ack.", append);
+                synchronized (state.getAckLock()) {
+                    // Acks must be sent in order. The only way to do this is by using a lock.
+                    long previousLastAcked = state.appendSuccessful(append.getEventNumber());
+                    if (previousLastAcked < append.getEventNumber()) {
+                        final DataAppended dataAppendedAck = new DataAppended(append.getRequestId(), append.getWriterId(), append.getEventNumber(),
+                                previousLastAcked, newWriteOffset);
+                        log.trace("Sending DataAppended : {}", dataAppendedAck);
+                        connection.send(dataAppendedAck);
```

Comment:
`connection.send(...)` does a `channel.writeAndFlush`. Did not understand the reason why we should we do this within the synchronized lock 🤔

Reply:
The client expects acks to arrive in the correct order (i.e., 1-5, 6-7,8-100). If they arrive out of order or have gaps it will freak out and close the connection.

This method is a callback from the `StreamSegmentStore.append`. While the Segment Store does issue acks in order, due to thread scheduling issues this callback may be invoked in arbitrary order. The code in `WriterState` does account for that, though.

The problem is that this callback is also subject to extreme race conditions. Since appends are pipelined to the store, the acks will not be serialized, but be issued in parallel. As such, more than one thread could be in this method. Thread A can invoke `writerState.appendSuccessful` and get ack range N1-N2 and Thread B can get range N2-N3. However Thread B could get to `connection.send` before thread A and, as such, send the acks out of order (N2-N3, followed by N1-N2).

The point of this sync block is to ensure that the acks are sent in order. 

If the client can be modified to handle acks out of order, given the pipelining guarantees of the store, we could remove that constraint from the client and simplify this code.